include/ScalarField.h

```diff
@@ -56,30 +100,120 @@ namespace CCCoreLib
+		inline ScalarType getMin() const { return m_offset + m_localMinVal; }
 		//! Returns the maximum value
-		inline ScalarType getMax() const { return m_maxVal; }
+		inline ScalarType getMax() const { return m_offset + m_localMaxVal; }
 
 		//! Fills the array with a particular value
-		inline void fill(ScalarType fillValue = 0) { if (empty()) resize(capacity(), fillValue); else std::fill(begin(), end(), fillValue); }
+		inline void fill(ScalarType fillValue = 0)
+		{
+			float fillValueF = 0.0f;
+			if (std::isfinite(fillValue))
+			{
+				if (fillValue == 0)
+				{
+					// special case: filling with zeros
+					// (it doesn't really give an idea of what the optimal offset is)
+					resetOffset();
+
+					//fillValueF = 0.0f; // already set
+				}
+				else
+				{
+					if (m_offsetHasBeenSet)
+					{
+						fillValueF = static_cast<float>(fillValue - m_offset);
+					}
+					else
+					{
+						// if the offset has not been set yet, we use the first finite value by default
+						setOffset(fillValue);
+
+						//fillValueF = 0.0f; // already set
+					}
+				}
+			}
+			else
+			{
+				// special case: filling with NaN or +/-inf values
+				// (it doesn't really give an idea of what the optimal offset is)
+				resetOffset();
+
+				fillValueF = static_cast<float>(fillValue); // NaN/-inf/+inf should be maintained
+			}
+
+			if (empty())
+			{
+				resize(capacity(), fillValueF);
+			}
+			else
+			{
+				std::fill(begin(), end(), fillValueF);
+			}
+		}
 
 		//! Reserves memory (no exception thrown)
 		CC_CORE_LIB_API bool reserveSafe(std::size_t count);
 		//! Resizes memory (no exception thrown)
 		CC_CORE_LIB_API bool resizeSafe(std::size_t count, bool initNewElements = false, ScalarType valueForNewElements = 0);
 
 		//Shortcuts (for backward compatibility)
-		inline ScalarType& getValue(std::size_t index) { return at(index); }
-		inline const ScalarType& getValue(std::size_t index) const { return at(index); }
-		inline void setValue(std::size_t index, ScalarType value) { at(index) = value; }
-		inline void addElement(ScalarType value) { push_back(value); }
+		inline ScalarType getValue(std::size_t index) const { return m_offset + (*this)[index]; }
+
+		inline float getLocalValue(std::size_t index) const { return (*this)[index]; }
+		inline void setLocalValue(std::size_t index, float value) { (*this)[index] = value; }
+		inline const float* getLocalValues() const { return data(); }
+
+		inline void setValue(std::size_t index, ScalarType value)
+		{
+			if (m_offsetHasBeenSet)
+			{
+				// use the already set offset
+				(*this)[index] = static_cast<float>(value - m_offset);
+			}
+			else if (std::isfinite(value))
+			{
+				// if the offset has not been set yet, we use the
+				// first finite value as offset by default
+				setOffset(value);
+				(*this)[index] = 0.0f;
+			}
+			else
+			{
+				// we can't set an offset
+				(*this)[index] = static_cast<float>(value); // NaN/-inf/+inf should be maintained
+			}
+		}
```

Comment:
Hum I wonder if this is ok in the general case:
If ScalarType is float and I have some big value store on a double, calling `setValue` will first cast my double into a float, so the loss of precision already happenned, its too late to apply the offset

The same thing goes for getValue which returns a ScalarType

Unlesss scalar type is no longer meant to be a float

Reply:
Indeed, ScalarType is double and should not be changed anymore (this would kill the purpose of all this code). I don't think we need to support the float case.

Reply:
In that case I think ScalarType should be removed (replaced with double) globally, or at least in these function, so its explicit that only double are expected